server/integration/endpoint/src/main/java/org/infinispan/server/endpoint/deployments/ProtoStreamMarshallerExtensionProcessor.java

```diff
@@ -0,0 +1,152 @@
+package org.infinispan.server.endpoint.deployments;
+
+import static org.infinispan.server.endpoint.EndpointLogger.ROOT_LOGGER;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.infinispan.commons.util.Util;
+import org.infinispan.protostream.BaseMarshaller;
+import org.infinispan.server.endpoint.Constants;
+import org.infinispan.server.endpoint.subsystem.ExtensionManagerService;
+import org.jboss.as.server.deployment.AttachmentKey;
+import org.jboss.as.server.deployment.Attachments;
+import org.jboss.as.server.deployment.DeploymentPhaseContext;
+import org.jboss.as.server.deployment.DeploymentUnit;
+import org.jboss.as.server.deployment.DeploymentUnitProcessor;
+import org.jboss.as.server.deployment.ServicesAttachment;
+import org.jboss.as.server.deployment.module.ResourceRoot;
+import org.jboss.modules.Module;
+import org.jboss.modules.ModuleClassLoader;
+import org.jboss.msc.service.Service;
+import org.jboss.msc.service.ServiceBuilder;
+import org.jboss.msc.service.ServiceController;
+import org.jboss.msc.service.ServiceName;
+import org.jboss.msc.service.StartContext;
+import org.jboss.msc.service.StartException;
+import org.jboss.msc.service.StopContext;
+import org.jboss.msc.value.InjectedValue;
+import org.jboss.vfs.VirtualFile;
+
+/**
+ * Processor for protobuf deployments, containing protostream marshallers and proto files.
+ *
+ * @since 9.3
+ */
+public class ProtoStreamMarshallerExtensionProcessor implements DeploymentUnitProcessor {
```

Comment:
Nice!

Reply:
Nice but dangerous like a hydrogen based Zeppelin :).  The unpredictable order of processing protos and marshallers will lead to chaos and cannot be completely controlled with dependencies declared via meta-inf manifest. Since the whole thing is based on deploying a jar (not a bad thing) I would go a bit further and give the user maximum programatic control. Rather than we do the auto-scanning of META-INF for proto files I would prefer to rely on ServiceLoader mechanism for the whole thing (not just the marshallers). Let's define an interface the user can implement that receives a SerializationContext and the user would be able to register marshallers and protos to it as he wishes. That way he can control the order and can handle errors, or at least fail with dignity, in a predictable way :)
```
interface ProtostreamSerializationContextInitializer {
 void init(SerializationContext ctx);
}
```
Ok, predictability and transparency solved, less mystery for the user.

Now the next question is why do we want a DeploymentUnitProcessor specifically for this. We can handle this quite easily by using the serviceloader mechanims to lookup ProtostreamSerializationContextInitializer implementations during  cache manager startup in the remote query module intializer handler.  How do these implementation classes coming from deployed jar become visible to remote query module intializer handler? We use the <modules> mechanism added in 9.1. That's one proposal to save us from adding one more DeploymentUnitProcessor that is probably not going to survive _The Second Coming_ of the server :)

Reply:
> The unpredictable order of processing protos and marshallers 

The problem that you are describing will happen when you deploy one jar with .proto only, and another with marshaller only; in the same jar it will work
 or am I missing something?

> why do we want a DeploymentUnitProcessor specifically for this
> We use the <modules> mechanism added in 9.1. That's one proposal to save us from adding > one more DeploymentUnitProcessor

On master, you can deploy a marshaller and it becomes the global event marshaller. So each time you are using filters/converters and protobuf caches in the server, you'd need to deploy your whole full-fledge ProtoStream marshaller, protofile, create your own serCtx, and so on. 

This PR is doing something similar, but it assumes .proto files are on META-INF/ and it reads and registers for you, but the big difference is now there is not more global event marshaller and user doesn't need to have its own ProtoStream marshaller, but it can extend the *existing* serCtx in the server, by adding extra messagemarshallers and protos. So it's easier to use.

>  That's one proposal to save us from adding one more DeploymentUnitProcessor

The new server will need to support deployed jars, we have too many of those scenarios already, so I am not worried about it

Your proposal could work, it gives more power but it's more legwork for the user, and this PR is already a great improvement over what we have now, so could we create a JIRA and postpone for 9.4?

Reply:
> We use the <modules> mechanism added in 9.1

btw, this also requires the ```<module>``` mechanism. Which means, the jar with the marshallers and .proto must be visible during the server startup, because marshalling involves entities and those entities need to be visible to the ProtoStream transcoder during starting

Reply:
Yes, such split jars will happen.

And also the order of registering marshallers may matter. Not for the most cases of BaseMarshaller but certainly for the generate ones.
 
Oh, speaking about the generated ones, what I don't like about proto files being regular files under meta-inf is that we cannot use the annotation based proto schema generator of protostream, which I would like people to use whenever possible. For generating the schema, and for generating the marshallers.

Also, protostream 5 will support a DSL with lambdas for defining marshallers, without implementing a class derived from BaseMarshaller. So looking up BaseMarshaller classes with ServiceLoader is not the whole story.

The more legwork for the user is implementing an interface with one method which is quite trivial and the code to be written there is what they already write for the client side. I'm not even suggesting they should copy-paste that code :) 

I like the fact there is no more global event marshaller, so I'm all in for most changes in this PR except this DeploymentUnitProcessor.

Reply:
There is also the question which of the many cache manager gets which protos and marshallers. My assumption is we have a good answer already: based on the contents of ```<modules>``` of each cache manager.   What gets listed under modules just needs to be a simple jar deployed under the deplouyments dir, not necessarily something that was specially processed by a DeploymentUnitProcessor.

Reply:
yes, that's how the PR works at the moment, I will simply remove the DUP and use service loader instead